src/codegen/inliner.lisp

```diff
@@ -1,313 +1,379 @@
+
+              ((null inner-rands)
+               (debug! ";; Inlining method to variable ~a" method-name)
+               (push method-name *functions-inlined*)
+               (ast:make-node-variable
+                :type (ast:node-type node)
+                :value method-name))
+
+              (t
+               (debug! ";; Inlining method to application ~a" method-name)
+               (push method-name *functions-inlined*)
+               (ast:make-node-application
+                :type (ast:node-type node)
+                :rator (ast:make-node-variable
+                        :type (tc:make-function-type*
+                               (mapcar #'ast:node-type inner-rands)
+                               (ast:node-type node))
+                        :value method-name)
+                :rands inner-rands))))
+          node))))
+
+(defun inline-direct-method (node env)
+  (declare (type ast:node-direct-application node)
+           (type tc:environment env)
+           (values ast:node &optional))
+
+  (unless *inline-methods-p*
+    (return-from inline-direct-method
+      node))
+
+  (let ((rands (ast:node-direct-application-rands node)))
+    (multiple-value-bind (dict inner-rands) (extract-dict rands)
+      (if dict
+          (let ((method-name (tc:lookup-method-inline env (ast:node-direct-application-rator node) dict :no-error t)))
+            (cond
+              ((null method-name)
+               node)
+
+              ((null inner-rands)
+               (debug! ";; Inlining direct method to variable ~a" method-name)
+               (push method-name *functions-inlined*)
+               (ast:make-node-variable
+                :type (ast:node-type node)
+                :value method-name))
+
+              (t
+               (debug! ";; Inlining direct method to application ~a" method-name)
+               (push method-name *functions-inlined*)
+               (ast:make-node-application
+                :type (ast:node-type node)
+                :rator (ast:make-node-variable
+                        :type (tc:make-function-type*
+                               (mapcar #'ast:node-type inner-rands)
+                               (ast:node-type node))
+                        :value method-name)
+                :rands inner-rands))))
+          node))))
+
+(defun inline-applications* (node env stack noinline-functions)
+  (declare (type ast:node node)
+           (type tc:environment env)
+           (type list stack)
+           (type parser:identifier-list noinline-functions)
+           (values ast:node &optional))
+
+  (traverse:traverse
+   node
+   (list
+    (traverse:action (:before ast:node-locally node)
+      (dolist (sym (ast:node-locally-noinline-functions node))
+        (push sym noinline-functions))
+      node)
+    (traverse:action (:before ast:node-application node)
+      (push (ast:node-rator-name node) stack)
+      node)
+    (traverse:action (:before ast:node-direct-application node)
```

Comment:
can we instead do a `:traverse` with a dynamic binding, so we don't need to do the pushes and pops manually?